### PR TITLE
Add query responses in logs

### PR DESCRIPTION
**Pull request by mateuszkulak**

### CODE REVIEW

File: src/Domain/LeanCode.CQRS.Execution/QueryLoggerElement.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        where AppContext : notnull, IPipelineContext
+    {
+        public readonly Serilog.ILogger logger = Serilog.Log.ForContext<QueryLoggerElement<AppContext, TObj, TRes>>();
+
+        public Task<TRes> ExecuteAsync(
+            AppContext ctx,
+            TObj input,
+            Func<AppContext, TObj, Task<TRes>> next)
+        {
+            logger.Information("Query executed with response {@Response}", input);
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/QueryLoggerElement.cs]: If you don't limit the `TObj` to be a `IQuery`, then this can be used for objects different than queries (e.g. operations, or even commands).

This is a very good approach, but let's not use `query` in names in that case.

File: src/Domain/LeanCode.CQRS.Execution/LeanCode.CQRS.Execution.csproj [hunk 2]
```diff
@@ -7,6 +7,7 @@
   <ItemGroup>
     <PackageReference Include="LeanCode.Contracts" />
     <PackageReference Include="Serilog" />
+    <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="7.0.0" />
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/LeanCode.CQRS.Execution.csproj]: The version is already pinned [here](https://github.com/leancodepl/corelibrary/blob/v7.0/Directory.Build.targets#L84-L85).
```suggestion
    <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" />
```

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace LeanCode.CQRS.Execution
+{
+    public class DevResponseLoggerElement<AppContext, TObj, TRes> : IPipelineElement<AppContext, TObj, TRes>
+        where AppContext : notnull, IPipelineContext
+    {
+        public readonly Serilog.ILogger logger = Serilog.Log.ForContext<DevResponseLoggerElement<AppContext, TObj, TRes>>();
+        public readonly IHostEnvironment environment;
+
+        DevResponseLoggerElement(IHostEnvironment env)
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]:
```suggestion
        public DevResponseLoggerElement(IHostEnvironment env)
```

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            logger = Serilog.Log.ForContext<DevResponseLoggerElement<AppContext, TObj, TRes>>();
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]:
```suggestion
```

You already initializing it in a field initializer.

You should have a separate ctor, that takes both `IHostEnvironment` and `ILogger` for tests only - but that one should be separate from the one that should be used most of the time.

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            environment = env;
+        }
+
+        public Task<TRes> ExecuteAsync(
+            AppContext ctx,
+            TObj input,
+            Func<AppContext, TObj, Task<TRes>> next)
+        {
+            if (environment.IsDevelopment())
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]:
```suggestion
            if (!environment.IsProduction())
```
Maybe? I'm not sure - we probably should ask on our internal Slack.

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs [hunk 6]
```diff
@@ -1,21 +1,25 @@
...
 using Microsoft.Extensions.Hosting;
+using Serilog;
 
 namespace LeanCode.CQRS.Execution
 {
     public class DevResponseLoggerElement<AppContext, TObj, TRes> : IPipelineElement<AppContext, TObj, TRes>
         where AppContext : notnull, IPipelineContext
     {
-        public readonly Serilog.ILogger logger = Serilog.Log.ForContext<DevResponseLoggerElement<AppContext, TObj, TRes>>();
+        public readonly ILogger logger;
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]: `private`

Comment (jakubfijalkowski) [reply src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]: The one below should also be private.

File: src/Domain/LeanCode.CQRS.Execution/ResponseLoggerElement.cs [hunk 7]
```diff
@@ -2,13 +2,24 @@
...
 using LeanCode.Pipelines;
+using Serilog;
 
 namespace LeanCode.CQRS.Execution
 {
     public class ResponseLoggerElement<AppContext, TObj, TRes> : IPipelineElement<AppContext, TObj, TRes>
         where AppContext : notnull, IPipelineContext
     {
-        public readonly Serilog.ILogger logger = Serilog.Log.ForContext<ResponseLoggerElement<AppContext, TObj, TRes>>();
+        public readonly ILogger logger;
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/ResponseLoggerElement.cs]: `private`

**Review by jakubfijalkowski — CHANGES_REQUESTED** [review]

File: test/Domain/LeanCode.CQRS.Tests/Execution/QueryExecutionTests.cs [hunk 8]
```diff
@@ -42,6 +46,30 @@ public async Task Correctly_routes_data_through_pipeline_elements()
             Assert.Equal(query, element.Data);
         }
 
+        [Fact]
+        public async Task Correctly_log_query_response_in_pipeline_element()
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/Execution/QueryExecutionTests.cs]: This is not the correct place to test this - you should have a separate test class for `ResponseLoggerElement` and `DevResponseLoggerElement` that tests just that. This class tests `QueryExecutionTests` and it has nothing to do with these two elements. Let's move that there. And also - we don't need whole pipeline here, we're just testing the single element so we can test it directly.

Comment (jakubfijalkowski) [reply test/Domain/LeanCode.CQRS.Tests/Execution/QueryExecutionTests.cs]: Same applies to the `OperationExecutionTests`.

**Review by jakubfijalkowski — CHANGES_REQUESTED** [review]

File: test/Domain/LeanCode.CQRS.Tests/Execution/OperationExecutionTests.cs [hunk 9]
```diff
@@ -1,4 +1,3 @@
-using System.Threading.Tasks;
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/Execution/OperationExecutionTests.cs]: Let's revert this change.

File: test/Domain/LeanCode.CQRS.Tests/Execution/QueryExecutionTests.cs [hunk 10]
```diff
@@ -1,4 +1,3 @@
-using System.Threading.Tasks;
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/Execution/QueryExecutionTests.cs]: And revert this one.

File: test/Domain/LeanCode.CQRS.Tests/BaseCQRSTests.cs [hunk 11]
```diff
@@ -42,6 +42,8 @@ public class BaseCQRSTests
             builder.RegisterType<SamplePipelineElement<ICommand, CommandResult>>().AsSelf().SingleInstance();
             builder.RegisterType<SamplePipelineElement<IQuery, object>>().AsSelf().SingleInstance();
             builder.RegisterType<SamplePipelineElement<IOperation, object>>().AsSelf().SingleInstance();
+            builder.RegisterType<ResponseLoggerElement<AppContext, object, object>>().AsSelf().SingleInstance();
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/BaseCQRSTests.cs]: This is not needed.

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+            var operationResponse = "Success!";
+            var nextFunctionResult = "Completed!";
+
+            var logger = new LoggerConfiguration().MinimumLevel
+                .Is(LogEventLevel.Verbose)
+                .WriteTo.Sink(fakeSink, LogEventLevel.Verbose)
+                .CreateLogger();
+
+            var element = Container.Resolve<ResponseLoggerElement<AppContext, object, object>>(new NamedParameter("logger", logger));
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]: You don't need the container. Just `new` the `ResponseLoggerElement`.

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        [Fact]
+        public async Task Correctly_log_operation_response_and_run_next()
+        {
+            Prepare();
+
+            var appCtx = new AppContext();
+            var fakeSink = new FakeSink();
+
+            var operationResponse = "Success!";
+            var nextFunctionResult = "Completed!";
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]:
```suggestion
            const OperationResponse = "Success!";
            const NextFunctionResult = "Completed!";
```

plus let's move it to the class level.

File: src/Domain/LeanCode.CQRS.Execution/ResponseLoggerElement.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/ResponseLoggerElement.cs]: Remember to fix the warnings in this file and in the `DevResponseLoggerElement`.

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        public Task<TRes> ExecuteAsync(
+            AppContext ctx,
+            TObj input,
+            Func<AppContext, TObj, Task<TRes>> next)
+        {
+            if (!environment.IsProduction())
+            {
+                logger.Information("Request executed with response {@Response}", input);
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]: You're logging the wrong thing - you should not log the input, but the response.

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+            var logger = new LoggerConfiguration().MinimumLevel
+                .Is(LogEventLevel.Verbose)
+                .WriteTo.Sink(fakeSink, LogEventLevel.Verbose)
+                .CreateLogger();
+
+            var element = Container.Resolve<ResponseLoggerElement<AppContext, object, object>>(new NamedParameter("logger", logger));
+
+            var result = await element.ExecuteAsync(appCtx, operationResponse, new Func<AppContext, object, Task<object>>((_, _) => Task.FromResult<object>(nextFunctionResult)));
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]:
```suggestion
            var result = await element.ExecuteAsync(appCtx, operationResponse, (ctx, obj) => Task.FromResult<object>(nextFunctionResult));
```

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+            Assert.Equal(1, fakeSink.Messages.Count());
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]:
```suggestion
            Assert.Single(fakeSink.Messages);
```

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+            Assert.Equal("Request executed with response \"Success!\"", fakeSink.Messages[0]);
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]: Don't check for the whole message, it does not matter. Check if the message contains what you expect (so the result of executing `next`).

Also, the test checks not what should be checked :P

**Review by jakubfijalkowski — CHANGES_REQUESTED** [review]

File: test/Domain/LeanCode.CQRS.Tests/Execution/OperationExecutionTests.cs (hunk 9 above)
Comment (mateuszkulak) [reply test/Domain/LeanCode.CQRS.Tests/Execution/OperationExecutionTests.cs]: VS Code says that this is unnecessary using directive and all tests are working without that so I dont understand why we need that

Comment (jakubfijalkowski) [reply test/Domain/LeanCode.CQRS.Tests/Execution/OperationExecutionTests.cs]: Because that is not the topic of this PR.

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs (hunk 15 above)
Comment (mateuszkulak) [reply src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]: As I understand it, in this situation input is the response

Comment (jakubfijalkowski) [reply src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]: The query handler will be executed when called `next`, not _before_. Elements are run _before_ or _after_ execution, where `before` is before calling `next`, and `after` is after calling `next`. `next` has the result, `input` is _the input_, so a query object.

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+namespace LeanCode.CQRS.Tests
+{
+    public class LoggingOperationResponsesTests : BaseCQRSTests
+    {
+        const string OperationResponse = "Success!";
+
+        [Fact]
+        public async Task Correctly_log_operation_response_and_run_next()
+        {
+            Prepare();
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]: Not needed.

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 20]
```diff
@@ -0,0 +1,83 @@
+using LeanCode.CQRS.Execution;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Hosting.Internal;
+using Serilog;
+using Serilog.Events;
+using Xunit;
+
+namespace LeanCode.CQRS.Tests
+{
+    public class LoggingOperationResponsesTests : BaseCQRSTests
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]:
```suggestion
    public class LoggingOperationResponsesTests
```

File: test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+            var result = await element.ExecuteAsync(appCtx, OperationResponse, (ctx, obj) => Task.FromResult<object>(obj));
+
+            Assert.Single(fakeSink.Messages);
+            Assert.Equal(OperationResponse, result);
+            Assert.Contains(result.ToString(), fakeSink.Messages[0]);
+        }
+
+        [Fact]
+        public async Task Does_not_log_operation_response_and_run_next_on_prod_env()
```
Comment (jakubfijalkowski) [inline test/Domain/LeanCode.CQRS.Tests/LoggingResponses/LoggingOperationResponsesTests.cs]: These are two tests - one that checks if the message is logged, one that checks if `next` is called.

File: src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs [hunk 22]
```diff
@@ -0,0 +1,38 @@
+using LeanCode.Pipelines;
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/DevResponseLoggerElement.cs]: There are still some warnings left (esp. regarding the generic parameter names).

File: test/Domain/LeanCode.CQRS.Tests/BaseCQRSTests.cs (hunk 11 above)
Comment (jakubfijalkowski) [reply test/Domain/LeanCode.CQRS.Tests/BaseCQRSTests.cs]: It's not resolved yet.

**Review by jakubfijalkowski — CHANGES_REQUESTED** [review]

**Comment by jakubfijalkowski** [comment]
Okay, very close now. TODO:
1. Helpers to register the elements, like this: https://github.com/leancodepl/corelibrary/blob/v7.0/src/Domain/LeanCode.CQRS.Security/CQRSSecurityElement.cs#L61-L82
2. Squash the commits, so that we the git history looks less cluttered.

File: src/Domain/LeanCode.CQRS.Execution/NonProductionResponseLoggerElement.cs [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+                logger.Information("Request executed with response {@Response}", result);
+            }
+
+            return result;
+        }
+    }
+
+    public static class LogResponsesnOnNonProductionPipelineBuilderExtensions
+    {
+        public static PipelineBuilder<TContext, TObj, TRes> LogRespones<TContext, TObj, TRes>(
```
Comment (jakubfijalkowski) [inline src/Domain/LeanCode.CQRS.Execution/NonProductionResponseLoggerElement.cs]: Let's rename this to sth like `LogResponesOnNonProduction` otherwise it clashes with the other `LogResponses`.

**Review by jakubfijalkowski — APPROVED** [review]
